### PR TITLE
Latest updates in DQM/TrackingMonitorSource for data/MC comparisons for the Tracking POG

### PR DESCRIPTION
**Pull request by dbruschi**
#### PR description:

This pull request contains the implementation of the latest distributions and plugins which are used by the Tracking POG for their data/MC comparisons. So there are basically new observables which the standard code doesn't contain but also different filters which allow to select events and/or tracks more specifically (for example Z->dilepton, ttbar).

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@dbruschi, 
I think the problem is that a `MuonTrackProducer` class already exists in `cmssw`: 

https://github.com/cms-sw/cmssw/blob/3c7284c8bf81079a0717558e20b76de261079e81/SimMuon/MCTruth/plugins/MuonTrackProducer.cc#L27

**Comment by mmusich** [comment]
@dbruschi can you squash all commits to avoid files in invalid state ? 
https://github.com/cms-sw/cmssw/pull/41787#issuecomment-1567049201

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by emanueleusai** [comment]
+1

File: DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      reco::TrackRef gtk = mu.globalTrack();
+      double chi2 = gtk->chi2();
+      double ndof = gtk->ndof();
+      double chbyndof = (ndof > 0) ? chi2 / ndof : 0;
+      if (chbyndof >= maxNormChi2_)
+        continue;
+
+      reco::TrackRef tk = mu.innerTrack();
+      double trkd0 = tk->d0();
+      double trkdz = tk->dz();
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc:77]:
```suggestion
```
No need to define, and assign, these two variables outside of the if scope below

File: DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      if (beamSpot.isValid()) {
+        trkd0 = -(tk->dxy(beamSpot->position()));
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc:79]:
```suggestion
        double trkd0 = -(tk->dxy(beamSpot->position()));
```

File: DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        if (std::fabs(trkd0) >= maxD0_)
+          continue;
+        trkdz = tk->dz(beamSpot->position());
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZtoMMMuonTrackProducer.cc:82]:
```suggestion
        double trkdz = tk->dz(beamSpot->position());
```

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+      reco::TrackRef gtk = mu.globalTrack();
+      double chi2 = gtk->chi2();
+      double ndof = gtk->ndof();
+      double chbyndof = (ndof > 0) ? chi2 / ndof : 0;
+      if (chbyndof >= maxNormChi2_)
+        continue;
+
+      reco::TrackRef tk = mu.innerTrack();
+      double trkd0 = tk->d0();
+      double trkdz = tk->dz();
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:215]:
```suggestion
```
(same as before)

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+      if (beamSpot.isValid()) {
+        trkd0 = -(tk->dxy(beamSpot->position()));
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:217]:
```suggestion
        double trkd0 = -(tk->dxy(beamSpot->position()));
```

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        if (std::fabs(trkd0) >= maxD0_)
+          continue;
+        trkdz = tk->dz(beamSpot->position());
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:220]:
```suggestion
        double trkdz = tk->dz(beamSpot->position());
```

File: DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#include <memory>
+#include <algorithm>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDProducer.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h:14]:
```suggestion
```
Already included in the .cc, not needed here

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+#include "DataFormats/EgammaCandidates/interface/Electron.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/HitPattern.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "TLorentzVector.h"
+//#include "RecoEgamma/ElectronIdentification/interface/CutBasedElectronID.h"
+#include "DataFormats/EgammaReco/interface/BasicCluster.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:12]:
```suggestion
```
Duplicated

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/Electron.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:4]:
```suggestion
```
Already in the .h

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 10]
```diff
@@ -0,0 +1,66 @@
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/Electron.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:3]:
```suggestion
```
These do not seem to be used anywhere here

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/HitPattern.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "TLorentzVector.h"
+//#include "RecoEgamma/ElectronIdentification/interface/CutBasedElectronID.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:10]:
```suggestion
```

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 12]
```diff
@@ -0,0 +1,66 @@
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/Electron.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/HitPattern.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "TLorentzVector.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:9]:
```suggestion
```
Is it needed?

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 13]
```diff
@@ -0,0 +1,66 @@
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/Electron.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/HitPattern.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:7]:
```suggestion
```
Is it needed?

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "TLorentzVector.h"
+//#include "RecoEgamma/ElectronIdentification/interface/CutBasedElectronID.h"
+#include "DataFormats/EgammaReco/interface/BasicCluster.h"
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:11]:
```suggestion
```
Is it needed?

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+void AlcaRecoTrackSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  std::unique_ptr<reco::TrackCollection> outputTColl(new reco::TrackCollection());
+
+  // Read Track collection
+  edm::Handle<reco::TrackCollection> tracks;
+  iEvent.getByToken(tracksToken_, tracks);
+
+  if (tracks.isValid()) {
+    for (auto const& trk : *tracks) {
+      //      std::cout << "alcareco track pt : " << trk.pt() << "  |  eta : " << trk.eta() << "  |  nhits : " << trk.hitPattern().numberOfAllHits(reco::HitPattern::TRACK_HITS) << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:43]:
```suggestion
```
Can this get removed?

File: DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+
+//                                                                        // class declaration                                                    //
+
+class AlcaRecoTrackSelector : public edm::one::EDProducer<> {
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h:23]: Why `one`? I have the impression that this could even be a `global` module

File: DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+public:
+  explicit AlcaRecoTrackSelector(const edm::ParameterSet&);
+  ~AlcaRecoTrackSelector() override = default;
+
+private:
+  void produce(edm::Event& iEvent, edm::EventSetup const& iSetup) override;
+
+  edm::ParameterSet parameters_;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/interface/AlcaRecoTrackSelector.h:31]:
```suggestion
```
This is probably not needed (see .cc)

File: DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+AlcaRecoTrackSelector::AlcaRecoTrackSelector(const edm::ParameterSet& ps)
+    : parameters_(ps),
+      tracksTag_(parameters_.getUntrackedParameter<edm::InputTag>("trackInputTag", edm::InputTag("generalTracks"))),
+      tracksToken_(consumes<reco::TrackCollection>(tracksTag_)),
+      ptmin_(parameters_.getUntrackedParameter<double>("ptmin", 0.)),
+      pmin_(parameters_.getUntrackedParameter<double>("pmin", 0.)),
+      etamin_(parameters_.getUntrackedParameter<double>("etamin", -4.)),
+      etamax_(parameters_.getUntrackedParameter<double>("etamax", 4.)),
+      nhits_(parameters_.getUntrackedParameter<uint32_t>("nhits", 1)) {
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/AlcaRecoTrackSelector.cc:30]: Why not, much more simply:
```suggestion
    : tracksTag_(ps.getUntrackedParameter<edm::InputTag>("trackInputTag", edm::InputTag("generalTracks"))),
      tracksToken_(consumes<reco::TrackCollection>(tracksTag_)),
      ptmin_(ps.getUntrackedParameter<double>("ptmin", 0.)),
      pmin_(ps.getUntrackedParameter<double>("pmin", 0.)),
      etamin_(ps.getUntrackedParameter<double>("etamin", -4.)),
      etamax_(ps.getUntrackedParameter<double>("etamax", 4.)),
      nhits_(ps.getUntrackedParameter<uint32_t>("nhits", 1)) {
```

**Review by perrotta — COMMENTED** [review]
This initial review takes into account the outcome of the static analyzer, and I further looked at only one (simple) class, the AlcaRecoTrackSelector. I'm convinced that the same cleanings and simplifications suggested for that class can be applied to most of the other classes introduced in this PR. Please adjust all them, and then the review can continue.

**Comment by perrotta** [comment]
please test

File: DQM/TrackingMonitorSource/interface/TrackMultiplicityFilter.h [hunk 19]
```diff
@@ -0,0 +1,39 @@
+#ifndef TRACKMULTIPLICITYFILTER_H
+#define TRACKMULTIPLICITYFILTER_H
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/interface/TrackMultiplicityFilter.h:2]: this doesn't comply with rule 4.1 [1] (that you follow for all the other plugins in this package).
Even better this could be merged with the plugin implementation itself, see rule 6.7 [1]

[1] https://cms-sw.github.io/cms_coding_rules.html

File: DQM/TrackingMonitorSource/interface/ZEEDetails.h [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "DQMServices/Core/interface/MonitorElement.h"
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+#include "HLTrigger/HLTcore/interface/HLTConfigProvider.h"
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/interface/ZEEDetails.h:22]: needed?

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 21]
```diff
@@ -112,13 +213,8 @@ class StandaloneTrackMonitor : public DQMEDAnalyzer {
...
   MonitorElement* vertexZposH_;
   MonitorElement* nVertexH_;
+  MonitorElement* nVtxH_;
 
-  MonitorElement* nPixBarrelH_;
-  MonitorElement* nPixEndcapH_;
-  MonitorElement* nStripTIBH_;
-  MonitorElement* nStripTOBH_;
-  MonitorElement* nStripTECH_;
-  MonitorElement* nStripTIDH_;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:121]: why are these removed?

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 22]
```diff
@@ -127,15 +223,93 @@ class StandaloneTrackMonitor : public DQMEDAnalyzer {
   MonitorElement* trueNIntH_;
 
   // Exclusive Quantities
-  MonitorElement* nHitsVspTH_;
-  MonitorElement* nHitsVsnVtxH_;
-  MonitorElement* nHitsVsEtaH_;
-  MonitorElement* nHitsVsCosThetaH_;
-  MonitorElement* nHitsVsPhiH_;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:134]: why are these removed?

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 23]
```diff
@@ -258,222 +488,687 @@ StandaloneTrackMonitor::StandaloneTrackMonitor(const edm::ParameterSet& ps)
...
+      ibook.book1D("hOffTrkClusWidthThin", "Off-track Cluster Width (Thin Sensor)", 20, -0.5, 19.5);
+  hOffTrkClusChargeThickH_ =
+      ibook.book1D("hOffTrkClusChargeThick", "Off-track Cluster Charge (Thick Sensor)", 100, 0, 1000);
+  hOffTrkClusWidthThickH_ =
+      ibook.book1D("hOffTrkClusWidthThick", "Off-track Cluster Width (Thick Sensor)", 20, -0.5, 19.5);
 }
 void StandaloneTrackMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup) {
-  // Get event setup (to get global transformation)
+  if (verbose_)
+    std::cout << "Begin StandaloneTrackMonitor" << std::endl;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1141]: `cout` is not thread safe. Please use either LogInfo or some other severity from the MessageLogging system.

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 24]
```diff
@@ -258,222 +488,687 @@ StandaloneTrackMonitor::StandaloneTrackMonitor(const edm::ParameterSet& ps)
...
 
-  siStripClusterInfo_.initEvent(iSetup);
+  // Get event setup (to get global transformation)
+  const TrackerGeometry& tkGeom = (*tkGeom_);
 
   // Primary vertex collection
   edm::Handle<reco::VertexCollection> vertexColl;
   iEvent.getByToken(vertexToken_, vertexColl);
+  if (!vertexColl.isValid()) {
+    std::cerr << "Error! Failed to get reco::Vertex Collection, for " << vertexTag_ << std::endl;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1152]: `cerr` is not thread safe. Please use either LogInfo or some other severity from the MessageLogging system.

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 25]
```diff
@@ -483,42 +1178,68 @@ void StandaloneTrackMonitor::analyze(edm::Event const& iEvent, edm::EventSetup c
...
+  // Get MVA and quality mask collections
   int ntracks = 0;
+
   if (tracks.isValid()) {
     edm::LogInfo("StandaloneTrackMonitor") << "Total # of Tracks: " << tracks->size();
     if (verbose_)
       edm::LogInfo("StandaloneTrackMonitor") << "Total # of Tracks: " << tracks->size();
     reco::Track::TrackQuality quality = reco::Track::qualityByName(trackQuality_);
+    if (verbose_)
+      std::cout << "Total # of Tracks: " << tracks->size() << std::endl;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1236]: ditto.

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc [hunk 26]
```diff
@@ -527,36 +1248,160 @@ void StandaloneTrackMonitor::analyze(edm::Event const& iEvent, edm::EventSetup c
...
-      double nValidTrackerHits = hitp.numberOfValidTrackerHits();
-      nHitsVsEtaH_->Fill(eta, nValidTrackerHits);
-      nHitsVsCosThetaH_->Fill(std::cos(theta), nValidTrackerHits);
-      nHitsVsPhiH_->Fill(phi, nValidTrackerHits);
-      nHitsVspTH_->Fill(pt, nValidTrackerHits);
-      nHitsVsnVtxH_->Fill(vertexColl->size(), nValidTrackerHits);
+      double nAllHits = hitp.numberOfAllHits(reco::HitPattern::TRACK_HITS);
+      double nAllTrackerHits = hitp.numberOfAllTrackerHits(reco::HitPattern::TRACK_HITS);
+
+      double trackdeltaR = 100000000000;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1254]:
```suggestion
      double trackdeltaR = std::numeric_limits<double>::max(); 
```
might be more elegant?

File: DQM/TrackingMonitorSource/plugins/TrackMultiplicityFilter.cc [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+  edm::Handle<reco::TrackCollection> tracks;
+  iEvent.getByToken(tracksToken_, tracks);
+  double count = std::count_if(tracks->begin(), tracks->end(), selector_);
+  std::cout << "count : " << count << std::endl;
+  std::cout << "nmin_ : " << nmin_ << std::endl;
+  std::cout << "tracks size : " << tracks->size() << std::endl;
+
+  pass = (count >= nmin_);
+
+  std::cout << "pass : " << pass << std::endl;
```
Comment (mmusich) [inline DQM/TrackingMonitorSource/plugins/TrackMultiplicityFilter.cc:38]: ditto.

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc (hunk 22 above)
Comment (dbruschi) [reply DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:134]: We changed them to nValidHitsvsnVtxH_, etc.

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc (hunk 23 above)
Comment (mmusich) [reply DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1196]: Is there a reason why this was not addressed?

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc (hunk 25 above)
Comment (mmusich) [reply DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1291]: Is there a reason why this was not addressed?

File: DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc (hunk 26 above)
Comment (mmusich) [reply DQM/TrackingMonitorSource/plugins/StandaloneTrackMonitor.cc:1309]: Is there a reason why this was not addressed?

**Comment by emanueleusai** [comment]
+1

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 28]
```diff
@@ -0,0 +1,405 @@
...
+      int nStripHits = hitp.numberOfValidStripHits();
+      if (nStripHits < minStripHits_)
+        continue;
+
+      // PF Isolation
+      reco::GsfElectron::PflowIsolationVariables pfIso = ele.pfIsolationVariables();
+      float absiso =
+          pfIso.sumChargedHadronPt + std::max(0.0, pfIso.sumNeutralHadronEt + pfIso.sumPhotonEt - 0.5 * pfIso.sumPUPt);
+      float eiso = absiso / (ele.pt());
+      if (eiso > maxIsoEle_)
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:142]:
```suggestion
      const float eiso =
          pfIso.sumChargedHadronPt + std::max(0.0, pfIso.sumNeutralHadronEt + pfIso.sumPhotonEt - 0.5 * pfIso.sumPUPt);
      if (eiso > maxIsoEle_ * ele.pt())
```

File: DQM/TrackingMonitorSource/plugins/ZEEDetails.cc [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+      int nStripHits = hitp.numberOfValidStripHits();
+      if (nStripHits < minStripHits_)
+        continue;
+
+      // DB corrected PF Isolation
+      reco::GsfElectron::PflowIsolationVariables pfIso = ele.pfIsolationVariables();
+      float absiso =
+          pfIso.sumChargedHadronPt + std::max(0.0, pfIso.sumNeutralHadronEt + pfIso.sumPhotonEt - 0.5 * pfIso.sumPUPt);
+      float eiso = absiso / (ele.pt());
+      if (eiso > maxIso_)
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZEEDetails.cc:151]:
```suggestion
      const float eiso =
          pfIso.sumChargedHadronPt + std::max(0.0, pfIso.sumNeutralHadronEt + pfIso.sumPhotonEt - 0.5 * pfIso.sumPUPt);
      if (eiso > maxIso_ * ele.pt())
```

File: DQM/TrackingMonitorSource/plugins/TrackMultiplicityFilter.cc [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+// ------------ method called on each new Event  ------------
+bool TrackMultiplicityFilter::filter(edm::StreamID iStream, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  bool pass = false;
+
+  edm::Handle<reco::TrackCollection> tracks;
+  iEvent.getByToken(tracksToken_, tracks);
+  double count = std::count_if(tracks->begin(), tracks->end(), selector_);
+  std::cout << "count : " << count << std::endl;
+  std::cout << "nmin_ : " << nmin_ << std::endl;
+  std::cout << "tracks size : " << tracks->size() << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/TrackMultiplicityFilter.cc:34]:
```suggestion
```
These look like comment left after your private debug: please remove (and in any case don't use thread unsafe `cout` in centrally run plugins)

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 31]
```diff
@@ -0,0 +1,405 @@
...
+      list_jets.push_back(lv_jets);
+    }
+    lj = list_jets.size();
+  }
+
+  edm::Handle<reco::JetTagCollection> bTagHandle;
+  iEvent.getByToken(bjetsToken_, bTagHandle);
+  const reco::JetTagCollection& bTags = *(bTagHandle.product());
+  std::vector<TLorentzVector> list_bjets;
+  std::cout << "nbjets " << bTags.size() << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:249]:
```suggestion
```
These look like comment left after your private debug: please remove (and in any case don't use thread unsafe `cout` in centrally run plugins)

File: DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc [hunk 32]
```diff
@@ -0,0 +1,405 @@
...
+      TLorentzVector lv_bjets;
+      lv_bjets.SetPtEtaPhiE(
+          bTags[bj].first->pt(), bTags[bj].first->eta(), bTags[bj].first->phi(), bTags[bj].first->energy());
+      if ((bTags[bj].second > btagFactor_) && (lv_bjets.Pt() > minPt_Jets_))
+        list_bjets.push_back(lv_bjets);
+    }
+    lbj = list_bjets.size();
+  }
+
+  std::cout << "le " << le << "\tlm " << lm << "\tlj " << lj << "\tlbj " << lbj << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ttbarEventSelector.cc:262]:
```suggestion
```
These look like comment left after your private debug: please remove (and in any case don't use thread unsafe `cout` in centrally run plugins)

File: DQM/TrackingMonitorSource/python/TrackingDataMCValidation_Standalone_cff.py [hunk 33]
```diff
@@ -1,44 +1,223 @@
...
+#   nbins = cms.int32 ( 100 ),
+#   xmin  = cms.double( 0.400),
+#   xmax  = cms.double( 0.600),
+# )
+
+# For MinBias
+standaloneTrackMonitorMC = standaloneTrackMonitor.clone(
+    puScaleFactorFile = cms.untracked.string("PileupScaleFactor_316060_wrt_nVertex_ZeroBias.root"),
+    doPUCorrection    = cms.untracked.bool(True),
+    isMC              = cms.untracked.bool(True)
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/python/TrackingDataMCValidation_Standalone_cff.py:110]:
```suggestion
    puScaleFactorFile = "PileupScaleFactor_316060_wrt_nVertex_ZeroBias.root",
    doPUCorrection    = True,
    isMC             =True
```
and remove explicit type declaration also from the other cloned modules (see e.g point 4 in https://twiki.cern.ch/twiki/bin/view/CMS/OpenRecoTasks#Review_and_optimization_of_the_p)

File: DQM/TrackingMonitorSource/python/TrackingDataMCValidation_Standalone_cff.py [hunk 34]
```diff
@@ -1,44 +1,223 @@
...
+    hltPathFilter
+    * selectedPrimaryVertices 
+#    * selectedMultiplicityTracks  # Use selectedMultiplicityTracks if needed nTracks > desired multiplicity
+#    * selectedAlcaRecoZBTracks
+    * selectedTracks
+    * standaloneTrackMonitorMC)
+# For ZtoEE
+standaloneTrackMonitorElec = standaloneTrackMonitor.clone(
+    folderName = cms.untracked.string("ElectronTracks"),
+    trackInputTag = cms.untracked.InputTag('electronTracks'),
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/python/TrackingDataMCValidation_Standalone_cff.py:129]:
```suggestion
    folderName = "ElectronTracks",
    trackInputTag = 'electronTracks',
```
etc.

File: DQM/TrackingMonitorSource/plugins/ZEEDetails.cc [hunk 35]
```diff
@@ -0,0 +1,223 @@
...
+      finalelectrons.push_back(ele);
+    }
+  } else {
+    edm::LogError("ElectronTrackProducer") << "Error >> Failed to get ElectronCollection for label: " << electronTag_;
+  }
+
+  edm::Handle<reco::VertexCollection> vertexColl;
+  iEvent.getByToken(vertexToken_, vertexColl);
+  if (!vertexColl.isValid()) {
+    std::cerr << "Error! Failed to get reco::Vertex Collection, for " << vertexTag_ << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZEEDetails.cc:170]:
```suggestion
```

File: DQM/TrackingMonitorSource/plugins/ZEEDetails.cc [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+    edm::LogError("DqmTrackStudy") << "Error! Failed to get reco::Vertex Collection, " << vertexTag_;
+  }
+  if (vertexColl->empty()) {
+    std::cerr << "No good vertex in the event!!" << std::endl;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/ZEEDetails.cc:174]:
```suggestion
```
either move to LogMessage's or remove

File: DQM/TrackingMonitorSource/plugins/TtbarTrackProducer.cc [hunk 37]
```diff
@@ -0,0 +1,139 @@
...
+      maxHOEEE_(ps.getUntrackedParameter<double>("maxHOEEB_", .10)),
+      maxSigmaiEiEEE_(ps.getUntrackedParameter<double>("maxSigmaiEiEEE", .03)),
+
+      // for Muon only
+      minChambers_(ps.getUntrackedParameter<uint32_t>("minChambers", 2)),
+      minMatches_(ps.getUntrackedParameter<uint32_t>("minMatches", 2)),
+      minMatchedStations_(ps.getUntrackedParameter<double>("minMatchedStations", 2)),
+
+      // for Jets only
+      minEtaHighest_Jets_(ps.getUntrackedParameter<double>("minEtaHighest_Jets", 2.4)),
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/plugins/TtbarTrackProducer.cc:51]: This does not look to be used

File: DQM/TrackingMonitorSource/interface/TtbarTrackProducer.h [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+  const double maxDeltaPhiInEE_;
+  const double maxDeltaEtaInEE_;
+  const double maxHOEEE_;
+  const double maxSigmaiEiEEE_;
+
+  const double minChambers_;
+  const double minMatches_;
+  const double minMatchedStations_;
+
+  const double minEtaHighest_Jets_;
```
Comment (perrotta) [inline DQM/TrackingMonitorSource/interface/TtbarTrackProducer.h:77]: Not used?

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by rappoccio** [comment]
+1
